### PR TITLE
Initial commit of log exporter

### PR DESCRIPTION
**Pull request by millerm**

### CODE REVIEW

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    public NabuLogExporter() {
+        this.logPosition = 0;
+        this.logFilePath = Path.of(System.getenv("NABU_TRACING_LOG_PATH"));
+        this.stateFilePath = Path.of(System.getenv("LOG_EXPORTER_STATE_PATH"));
+
+        try {
+            this.readState();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Comment (millerm) [inline nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java:153]: TODO: move to an initialize() function

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+import java.util.concurrent.Executors;
+
+public class NabuLogExporter implements LogExporter {
+    private static final Logger LOG = Logger.getGlobal();
+
+    private static final int SLEEP_TIME_MILLISECONDS = 1;
+
+    private static final int MAX_LOG_FILE_SIZE = 200 * 1024 * 1024;
+
+    private final Path LOG_PATH_DIR;
```
Comment (sonudoo) [inline nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java]: **LOG_PATH_DIR** is not a static final field, can we have camelCase for it?

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            LOG.severe("Error watching directory: " + e.getMessage());
+        }
+    }
+
+    private static class LogTask implements Runnable {
+        private final Path filePath;
+        private final String stateFilePath;
+        long filePointer;
+
+        public LogTask(Path filePath, Path stateFilePath) {
```
Comment (sonudoo) [inline nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java]: Should `stateFilePath` be called `stateDirPath`?

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                    "\"operationName\":\"" + parts[5] + "\"," +
+                    "\"references\":[]," +
+                    "\"startTimeUnixNano\":\"" + "1716525641" + "\"," +
+                    "\"endTimeUnixNano\":\"" + "1716525642" + "\"," +
+                    "\"kind\":2" +
+                    "}";
+
+        }
+
+        private static void handleLog(String line) {
```
Comment (sonudoo) [inline nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java:160]: Consider adding some commentary description to each methods for readability.

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+                while (true) {
+                    // NOTE(@millerm): Can switch to read() to read a specific amount
+                    // of bytes if necessary
+                    logEntry = reader.readLine();
+
+                    if (logEntry != null) {
+                        LOG.info("New log: " + logEntry);
+
+                        handleLog(logEntry);
+                        saveState(reader.getFilePointer());
```
Comment (sonudoo) [inline nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java:211]: Is there a harm in pushing duplicate logs to the backend? May be we can saveState() in a background thread?

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+                    } else {
+                        Thread.sleep(SLEEP_TIME_MILLISECONDS);
```
Comment (sonudoo) [inline nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java]: Should `SLEEP_TIME_MILLISECONDS` be called LOG_EXPORT_POLL_DELAY_MILLISECONDS?

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+                    }
+                }
+            } catch (IOException | InterruptedException e) {
+                LOG.severe("Error reading log file: " + e.getMessage());
+                Thread.currentThread().interrupt();
+            }
+        }
+
+        private void saveState(long position) {
+            try (BufferedWriter writer = Files.newBufferedWriter(Path.of(stateFilePath), StandardOpenOption.CREATE,
```
Comment (sonudoo) [inline nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java]: Opening the file again and again is expensive. Consider PWrite() may be?

**Review by sonudoo — APPROVED** [review]
I had a few nitpicks. It would be good to address them before merge. Thanks!

**Review by sonudoo — COMMENTED** [review]
Can you rename `nabu-log-exporter` to `daemon`?

It is already in the `nabu-tracer` repo so the name should be clear.

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java (hunk 2 above)
Comment (millerm) [reply nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java]: Good catch!

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java (hunk 3 above)
Comment (millerm) [reply nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java]: I think that's a better name, great suggestion.

File: nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+            String[] parts = line.split(" ", 7);
+
+            // TODO: better parse
+            return "{" +
+                    "\"traceId\":\"" + parts[0] + "\"," +
+                    "\"spanId\":\"" + parts[1] + "\"," +
+                    "\"operationName\":\"" + parts[5] + "\"," +
+                    "\"references\":[]," +
+                    "\"startTimeUnixNano\":\"" + "1716525641" + "\"," +
+                    "\"endTimeUnixNano\":\"" + "1716525642" + "\"," +
```
Comment (goorui) [inline nabu-log-exporter/src/main/java/com/github/millerm/NabuLogExporter.java:154]: This part prob needs changing now that the daemon process is sending timestamp-based logs

**Pull request merged**